mathjax3-ts/input/tex/NodeFactory.ts

```diff
@@ -0,0 +1,195 @@
+ *  
+ *  Implements the TeX InputJax that reads mathematics in
+ *  TeX and LaTeX format and converts it to the MML ElementJax
+ *  internal format.
+ *
+ *  ---------------------------------------------------------------------
+ *  
+ *  Copyright (c) 2009-2018 The MathJax Consortium
+ * 
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+/**
+ * @fileoverview Node factory for creating MmlNodes. This allows extension
+ *     packages to add node constructors or overwrite existing ones.
+ *
+ * @author v.sorge@mathjax.org (Volker Sorge)
+ */
+
+import {TextNode, MmlNode, AbstractMmlNode, AbstractMmlEmptyNode} from '../../core/MmlTree/MmlNode.js';
+import {MmlMo} from '../../core/MmlTree/MmlNodes/mo.js';
+import {Property, PropertyList} from '../../core/Tree/Node.js';
+import {MmlFactory} from '../../core/MmlTree/MmlFactory.js';
+import {Args} from './Types.js';
+import {OperatorDef} from '../../core/MmlTree/OperatorDictionary.js';
+import TexParser from './TexParser.js';
+import ParseOptions from './ParseOptions.js';
+import NodeUtil from './NodeUtil.js';
+
+
+export type NodeFactoryMethod = (factory: NodeFactory, kind: string, ...rest: any[]) => MmlNode;
+
+export class NodeFactory {
+
+  /**
+   * Parser configuration that can be used for passes information between node methods.
+   * @type {ParseOption}
+   */
+  public configuration: ParseOptions;
+
+
+  /**
+   * The external node factory.
+   * @type {MmlFactory}
+   */
+  protected mmlFactory: MmlFactory = new MmlFactory();
+
+
+  /**
+   * The factory table populated with some default methods.
+   */
+  private factory: Map<string, NodeFactoryMethod> =
+    new Map([
+      ['node', NodeFactory.createNode],
+      ['token', NodeFactory.createToken],
+      ['text', NodeFactory.createText],
+      ['error', NodeFactory.createError],
+    ]);
+
+
+  /**
+   * Default node generation function.
+   * @param {NodeFactory} factory The current node factory.
+   * @param {string} kind The type of node to create.
+   * @param {MmlNode[]} children Its children.
+   * @param {any} def Its properties.
+   * @param {TextNode=} text An optional text node if this is a token.
+   * @return {MmlNode} The newly created Mml node.
+   */
+  public static createNode(factory: NodeFactory, kind: string,
+                           children: MmlNode[], def: any,
+                           text?: TextNode): MmlNode {
+    const node = factory.mmlFactory.create(kind, {}, []);
```

Comment:
Well you caught me there... The order is due to the fact `createNode` will always have children but not necessarily `def`, otherwise one would normally use `createToken`. So `def` was supposed to be optional. But for some reason I never really enforced this during code cleanup.

I agree in general that we should try to be as consistent as possible with the argument order. In this case, however, I will change that order (i.e., children before properties), as it will allow me to get rid of `{}` arguments and actually shorten the code.